openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py

```diff
@@ -0,0 +1,444 @@
+        self.clip_test = clip_test
+        self.round = 0    # starting round
+        self.aggregated_model_accuracy = None
+        self.privacy_accountant = RDPAccountant()
+        config = parse_config(config_path)
+
+        if 'differential_privacy' not in config:
+            self.dp_params = None
+        else:
+            self.dp_params = config["differential_privacy"]
+            print("Here are dp_params: {}".format(self.dp_params))
+            validate_dp_params(self.dp_params)
+            
+    @aggregator
+    def start(self):
+        print(f'Performing initialization for model')
+        self.collaborators = self.runtime.collaborators
+        self.private = 10
+
+        if self.dp_params is None:
+            self.round_collaborators = self.collaborators
+            self.optimizers = {collaborator_name: default_optimizer(self.model) for collaborator_name in self.collaborator_names}
+
+        else:
+            self.optimizers = {collaborator_name: ClipOptimizer(base_optimizer=default_optimizer(self.model),
+                                                   device=self.device, 
+                                                   clip_norm=self.dp_params['clip_norm'], 
+                                                   clip_freq=self.dp_params['clip_frequency']) for collaborator_name in self.collaborator_names}
+            self.sample_rate = self.dp_params['sample_rate']
+            global_data_loader = DataLoader(self.collaborators, batch_size=int(self.sample_rate * float(len(self.collaborators))))
+            dp_data_loader = DPDataLoader.from_data_loader(global_data_loader, distributed=False)
+            collaborator_batch = []
+            batch_sizes = []
+            for cols in dp_data_loader:
+                batch_sizes.append(len(cols))
+                collaborator_batch.append(cols)
+            self.round_collaborators = collaborator_batch[0]
+
+        if len(self.round_collaborators) !=0:
+            if not isinstance(self.round_collaborators[0], torch.Tensor):
+                print(20*"#")
+                print(f'Round {self.round}...')
+                print("Batch sizes sampled:", batch_sizes)
+                print(f"Collaborators patricipated in Round: {self.round}", self.round_collaborators)
+                print(20*"#") 
+                self.next(self.aggregated_model_validation, foreach='round_collaborators', exclude=['private'])
+            else:
+                print(f'No collaborator selected for training at Round: {self.round}')
+                self.next(self.check_round_completion)
+        else:
+            print(f'No collaborator selected for training at Round: {self.round}')
+            self.next(self.check_round_completion)
+
+    # @collaborator                     # Uncomment this if you don't have GPU in the machine and want this application ro run on CPU instead 
+    @collaborator(num_gpus=1)           # Assuming GPU(s) is available in the machine
```

Comment:
@mansishr: as @psfoley suggested, I kept `backend='single_process'`  by default as the memory leak in Ray is not completely cured yet. @psfoley will add a comment to demonstrate how to enable Ray backend as follows:
`local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='ray')`

Reply:
In the case of the `single_process` backend for the `LocalRuntime`, all models will be sent to the same device. This should be okay because with the single_process backend, only one GPU will be able to used simultaneously for this example. 

If there are multiple GPUs and sufficient memory in the system to take advantage of concurrent collaborator training, (`local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='ray')`), Ray will automatically set  `CUDA_VISIBLE_DEVICES` within the task, and each collaborator will be able to use their own dedicated GPU without the user needing to make any change to the workflow definition. 

It's worth noting though that setting `num_gpus=1` is only honored by the Ray backend - the `single_process` backend will just ignore this. I'll add a warning / info message when these placement decorator arguments are ignored for backends where they are not supported